// @neo4j/graphql/dist/classes/Node.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const camelcase_1 = __importDefault(require("camelcase"));
const pluralize_1 = __importDefault(require("pluralize"));
class Node {
    constructor(input) {
        this.name = input.name;
        this.relationFields = input.relationFields;
        this.connectionFields = input.connectionFields;
        this.cypherFields = input.cypherFields;
        this.primitiveFields = input.primitiveFields;
        this.scalarFields = input.scalarFields;
        this.enumFields = input.enumFields;
        this.otherDirectives = input.otherDirectives;
        this.unionFields = input.unionFields;
        this.interfaceFields = input.interfaceFields;
        this.interfaces = input.interfaces;
        this.objectFields = input.objectFields;
        this.temporalFields = input.temporalFields;
        this.pointFields = input.pointFields;
        this.ignoredFields = input.ignoredFields;
        this.exclude = input.exclude;
        this.nodeDirective = input.nodeDirective;
        this.auth = input.auth;
        this.description = input.description;
        this.authableFields = [
            ...input.primitiveFields,
            ...input.scalarFields,
            ...input.enumFields,
            ...input.unionFields,
            ...input.objectFields,
            ...input.temporalFields,
            ...input.pointFields,
            ...input.cypherFields,
        ];
        this.mutableFields = [
            ...input.temporalFields,
            ...input.enumFields,
            ...input.objectFields,
            ...input.scalarFields,
            ...input.primitiveFields,
            ...input.interfaceFields,
            ...input.objectFields,
            ...input.unionFields,
            ...input.pointFields,
        ];
    }
    get labelString() {
        var _a;
        return ((_a = this.nodeDirective) === null || _a === void 0 ? void 0 : _a.getLabelsString(this.name)) || `:${this.name}`;
    }
    get labels() {
        var _a;
        return ((_a = this.nodeDirective) === null || _a === void 0 ? void 0 : _a.getLabels(this.name)) || [this.name];
    }
    getPlural(options) {
        var _a;
        // camelCase is optional in this case to maintain backward compatibility
        if ((_a = this.nodeDirective) === null || _a === void 0 ? void 0 : _a.plural) {
            return options.camelCase ? camelcase_1.default(this.nodeDirective.plural) : this.nodeDirective.plural;
        }
        return pluralize_1.default(options.camelCase ? camelcase_1.default(this.name) : this.name);
    }
}
exports.default = Node;
