// @neo4j/graphql/dist/translate/projection/elements/create-datetime-element.test.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const create_datetime_element_1 = __importDefault(require("./create-datetime-element"));
describe("createDatetimeElement", () => {
    test("returns projection element for single datetime value", () => {
        const resolveTree = {
            name: "datetime",
            alias: "datetime",
            args: {},
            fieldsByTypeName: {},
        };
        const field = {
            // @ts-ignore
            typeMeta: {
                name: "DateTime",
            },
        };
        const element = create_datetime_element_1.default({
            resolveTree,
            field,
            variable: "this",
        });
        expect(element).toEqual('datetime: apoc.date.convertFormat(toString(this.datetime), "iso_zoned_date_time", "iso_offset_date_time")');
    });
    test("returns projection element for array of datetime values", () => {
        const resolveTree = {
            name: "datetimes",
            alias: "datetimes",
            args: {},
            fieldsByTypeName: {},
        };
        const field = {
            // @ts-ignore
            typeMeta: {
                name: "DateTime",
                array: true,
            },
        };
        const element = create_datetime_element_1.default({
            resolveTree,
            field,
            variable: "this",
        });
        expect(element).toEqual('datetimes: [ dt in this.datetimes | apoc.date.convertFormat(toString(dt), "iso_zoned_date_time", "iso_offset_date_time") ]');
    });
});
