// @neo4j/graphql/dist/translate/create-create-and-params.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const create_connect_and_params_1 = __importDefault(require("./create-connect-and-params"));
const create_auth_and_params_1 = __importDefault(require("./create-auth-and-params"));
const constants_1 = require("../constants");
const create_set_relationship_properties_and_params_1 = __importDefault(require("./create-set-relationship-properties-and-params"));
const map_to_db_property_1 = __importDefault(require("../utils/map-to-db-property"));
function createCreateAndParams({ input, varName, node, context, withVars, insideDoWhen, }) {
    function reducer(res, [key, value]) {
        const _varName = `${varName}_${key}`;
        const relationField = node.relationFields.find((x) => key === x.fieldName);
        const primitiveField = node.primitiveFields.find((x) => key === x.fieldName);
        const pointField = node.pointFields.find((x) => key === x.fieldName);
        const dbFieldName = map_to_db_property_1.default(node, key);
        if (relationField) {
            const refNodes = [];
            // let unionTypeName = "";
            if (relationField.union) {
                // [unionTypeName] = key.split(`${relationField.fieldName}_`).join("").split("_");
                Object.keys(value).forEach((unionTypeName) => {
                    refNodes.push(context.neoSchema.nodes.find((x) => x.name === unionTypeName));
                });
                // refNode = context.neoSchema.nodes.find((x) => x.name === unionTypeName) as Node;
            }
            else {
                refNodes.push(context.neoSchema.nodes.find((x) => x.name === relationField.typeMeta.name));
            }
            refNodes.forEach((refNode) => {
                const v = relationField.union ? value[refNode.name] : value;
                const unionTypeName = relationField.union ? refNode.name : "";
                if (v.create) {
                    const creates = relationField.typeMeta.array ? v.create : [v.create];
                    creates.forEach((create, index) => {
                        var _a;
                        res.creates.push(`\nWITH ${withVars.join(", ")}`);
                        const baseName = `${_varName}${relationField.union ? "_" : ""}${unionTypeName}${index}`;
                        const nodeName = `${baseName}_node`;
                        const propertiesName = `${baseName}_relationship`;
                        const recurse = createCreateAndParams({
                            input: create.node,
                            context,
                            node: refNode,
                            varName: nodeName,
                            withVars: [...withVars, nodeName],
                        });
                        res.creates.push(recurse[0]);
                        res.params = Object.assign(Object.assign({}, res.params), recurse[1]);
                        const inStr = relationField.direction === "IN" ? "<-" : "-";
                        const outStr = relationField.direction === "OUT" ? "->" : "-";
                        const relTypeStr = `[${relationField.properties ? propertiesName : ""}:${relationField.type}]`;
                        res.creates.push(`MERGE (${varName})${inStr}${relTypeStr}${outStr}(${nodeName})`);
                        if (relationField.properties) {
                            const relationship = context.neoSchema.relationships.find((x) => x.properties === relationField.properties);
                            const setA = create_set_relationship_properties_and_params_1.default({
                                properties: (_a = create.edge) !== null && _a !== void 0 ? _a : {},
                                varName: propertiesName,
                                relationship,
                                operation: "CREATE",
                            });
                            res.creates.push(setA[0]);
                            res.params = Object.assign(Object.assign({}, res.params), setA[1]);
                        }
                    });
                }
                if (v.connect) {
                    const connectAndParams = create_connect_and_params_1.default({
                        withVars,
                        value: v.connect,
                        varName: `${_varName}${relationField.union ? "_" : ""}${unionTypeName}_connect`,
                        parentVar: varName,
                        relationField,
                        context,
                        refNode,
                        labelOverride: unionTypeName,
                        parentNode: node,
                        fromCreate: true,
                    });
                    res.creates.push(connectAndParams[0]);
                    res.params = Object.assign(Object.assign({}, res.params), connectAndParams[1]);
                }
            });
            return res;
        }
        if (primitiveField === null || primitiveField === void 0 ? void 0 : primitiveField.auth) {
            const authAndParams = create_auth_and_params_1.default({
                entity: primitiveField,
                operation: "CREATE",
                context,
                bind: { parentNode: node, varName, chainStr: _varName },
                escapeQuotes: Boolean(insideDoWhen),
            });
            if (authAndParams[0]) {
                if (!res.meta) {
                    res.meta = { authStrs: [] };
                }
                res.meta.authStrs.push(authAndParams[0]);
                res.params = Object.assign(Object.assign({}, res.params), authAndParams[1]);
            }
        }
        if (pointField) {
            if (pointField.typeMeta.array) {
                res.creates.push(`SET ${varName}.${dbFieldName} = [p in $${_varName} | point(p)]`);
            }
            else {
                res.creates.push(`SET ${varName}.${dbFieldName} = point($${_varName})`);
            }
            res.params[_varName] = value;
            return res;
        }
        res.creates.push(`SET ${varName}.${dbFieldName} = $${_varName}`);
        res.params[_varName] = value;
        return res;
    }
    const labels = node.labelString;
    const initial = [`CREATE (${varName}${labels})`];
    const timestampedFields = node.temporalFields.filter((x) => { var _a; return ["DateTime", "Time"].includes(x.typeMeta.name) && ((_a = x.timestamps) === null || _a === void 0 ? void 0 : _a.includes("CREATE")); });
    timestampedFields.forEach((field) => {
        // DateTime -> datetime(); Time -> time()
        initial.push(`SET ${varName}.${field.dbPropertyName} = ${field.typeMeta.name.toLowerCase()}()`);
    });
    const autogeneratedIdFields = node.primitiveFields.filter((x) => x.autogenerate);
    autogeneratedIdFields.forEach((f) => {
        initial.push(`SET ${varName}.${f.dbPropertyName} = randomUUID()`);
    });
    // eslint-disable-next-line prefer-const
    let { creates, params, meta } = Object.entries(input).reduce(reducer, {
        creates: initial,
        params: {},
    });
    const forbiddenString = insideDoWhen ? `\\"${constants_1.AUTH_FORBIDDEN_ERROR}\\"` : `"${constants_1.AUTH_FORBIDDEN_ERROR}"`;
    if (node.auth) {
        const bindAndParams = create_auth_and_params_1.default({
            entity: node,
            operation: "CREATE",
            context,
            bind: { parentNode: node, varName },
            escapeQuotes: Boolean(insideDoWhen),
        });
        if (bindAndParams[0]) {
            creates.push(`WITH ${withVars.join(", ")}`);
            creates.push(`CALL apoc.util.validate(NOT(${bindAndParams[0]}), ${forbiddenString}, [0])`);
            params = Object.assign(Object.assign({}, params), bindAndParams[1]);
        }
    }
    if (meta === null || meta === void 0 ? void 0 : meta.authStrs.length) {
        creates.push(`WITH ${withVars.join(", ")}`);
        creates.push(`CALL apoc.util.validate(NOT(${meta.authStrs.join(" AND ")}), ${forbiddenString}, [0])`);
    }
    return [creates.join("\n"), params];
}
exports.default = createCreateAndParams;
