// @neo4j/graphql/dist/translate/translate-update.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const create_where_and_params_1 = __importDefault(require("./create-where-and-params"));
const create_projection_and_params_1 = __importDefault(require("./create-projection-and-params"));
const create_create_and_params_1 = __importDefault(require("./create-create-and-params"));
const create_auth_and_params_1 = __importDefault(require("./create-auth-and-params"));
const create_update_and_params_1 = __importDefault(require("./create-update-and-params"));
const create_connect_and_params_1 = __importDefault(require("./create-connect-and-params"));
const create_disconnect_and_params_1 = __importDefault(require("./create-disconnect-and-params"));
const constants_1 = require("../constants");
const create_delete_and_params_1 = __importDefault(require("./create-delete-and-params"));
const create_connection_and_params_1 = __importDefault(require("./connection/create-connection-and-params"));
const create_set_relationship_properties_and_params_1 = __importDefault(require("./create-set-relationship-properties-and-params"));
function translateUpdate({ node, context }) {
    var _a, _b, _c, _d;
    const { resolveTree } = context;
    const whereInput = resolveTree.args.where;
    const updateInput = resolveTree.args.update;
    const connectInput = resolveTree.args.connect;
    const disconnectInput = resolveTree.args.disconnect;
    const createInput = resolveTree.args.create;
    const deleteInput = resolveTree.args.delete;
    const varName = "this";
    const labels = node.labelString;
    const matchStr = `MATCH (${varName}${labels})`;
    let whereStr = "";
    let updateStr = "";
    const connectStrs = [];
    const disconnectStrs = [];
    const createStrs = [];
    let deleteStr = "";
    let projAuth = "";
    let projStr = "";
    let cypherParams = {};
    const whereStrs = [];
    const connectionStrs = [];
    let updateArgs = {};
    // Due to potential aliasing of returned object in response we look through fields of UpdateMutationResponse
    // and find field where field.name ~ node.name which exists by construction
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const { fieldsByTypeName } = Object.values(resolveTree.fieldsByTypeName[`Update${node.getPlural({ camelCase: false })}MutationResponse`]).find((field) => field.name === node.getPlural({ camelCase: true }));
    if (whereInput) {
        const where = create_where_and_params_1.default({
            whereInput,
            varName,
            node,
            context,
            recursing: true,
        });
        if (where[0]) {
            whereStrs.push(where[0]);
            cypherParams = Object.assign(Object.assign({}, cypherParams), where[1]);
        }
    }
    const whereAuth = create_auth_and_params_1.default({
        operation: "UPDATE",
        entity: node,
        context,
        where: { varName, node },
    });
    if (whereAuth[0]) {
        whereStrs.push(whereAuth[0]);
        cypherParams = Object.assign(Object.assign({}, cypherParams), whereAuth[1]);
    }
    if (whereStrs.length) {
        whereStr = `WHERE ${whereStrs.join(" AND ")}`;
    }
    if (updateInput) {
        const updateAndParams = create_update_and_params_1.default({
            context,
            node,
            updateInput,
            varName,
            parentVar: varName,
            withVars: [varName],
            parameterPrefix: `${resolveTree.name}.args.update`,
        });
        [updateStr] = updateAndParams;
        cypherParams = Object.assign(Object.assign({}, cypherParams), updateAndParams[1]);
        updateArgs = Object.assign(Object.assign({}, updateArgs), (updateStr.includes(resolveTree.name) ? { update: updateInput } : {}));
    }
    if (disconnectInput) {
        Object.entries(disconnectInput).forEach((entry) => {
            const relationField = node.relationFields.find((x) => x.fieldName === entry[0]);
            const refNodes = [];
            if (relationField.union) {
                Object.keys(entry[1]).forEach((unionTypeName) => {
                    refNodes.push(context.neoSchema.nodes.find((x) => x.name === unionTypeName));
                });
            }
            else {
                refNodes.push(context.neoSchema.nodes.find((x) => x.name === relationField.typeMeta.name));
            }
            refNodes.forEach((refNode) => {
                const disconnectAndParams = create_disconnect_and_params_1.default({
                    context,
                    parentVar: varName,
                    refNode,
                    relationField,
                    value: relationField.union ? entry[1][refNode.name] : entry[1],
                    varName: `${varName}_disconnect_${entry[0]}${relationField.union ? `_${refNode.name}` : ""}`,
                    withVars: [varName],
                    parentNode: node,
                    parameterPrefix: `${resolveTree.name}.args.disconnect.${entry[0]}${relationField.union ? `.${refNode.name}` : ""}`,
                    labelOverride: relationField.union ? refNode.name : "",
                });
                disconnectStrs.push(disconnectAndParams[0]);
                cypherParams = Object.assign(Object.assign({}, cypherParams), disconnectAndParams[1]);
            });
        });
        updateArgs = Object.assign(Object.assign({}, updateArgs), { disconnect: disconnectInput });
    }
    if (connectInput) {
        Object.entries(connectInput).forEach((entry) => {
            const relationField = node.relationFields.find((x) => entry[0] === x.fieldName);
            const refNodes = [];
            if (relationField.union) {
                Object.keys(entry[1]).forEach((unionTypeName) => {
                    refNodes.push(context.neoSchema.nodes.find((x) => x.name === unionTypeName));
                });
            }
            else {
                refNodes.push(context.neoSchema.nodes.find((x) => x.name === relationField.typeMeta.name));
            }
            refNodes.forEach((refNode) => {
                const connectAndParams = create_connect_and_params_1.default({
                    context,
                    parentVar: varName,
                    refNode,
                    relationField,
                    value: relationField.union ? entry[1][refNode.name] : entry[1],
                    varName: `${varName}_connect_${entry[0]}${relationField.union ? `_${refNode.name}` : ""}`,
                    withVars: [varName],
                    parentNode: node,
                    labelOverride: relationField.union ? refNode.name : "",
                });
                connectStrs.push(connectAndParams[0]);
                cypherParams = Object.assign(Object.assign({}, cypherParams), connectAndParams[1]);
            });
        });
    }
    if (createInput) {
        Object.entries(createInput).forEach((entry) => {
            const relationField = node.relationFields.find((x) => entry[0] === x.fieldName);
            const refNodes = [];
            if (relationField.union) {
                Object.keys(entry[1]).forEach((unionTypeName) => {
                    refNodes.push(context.neoSchema.nodes.find((x) => x.name === unionTypeName));
                });
            }
            else {
                refNodes.push(context.neoSchema.nodes.find((x) => x.name === relationField.typeMeta.name));
            }
            const inStr = relationField.direction === "IN" ? "<-" : "-";
            const outStr = relationField.direction === "OUT" ? "->" : "-";
            refNodes.forEach((refNode) => {
                const v = relationField.union ? entry[1][refNode.name] : entry[1];
                const creates = relationField.typeMeta.array ? v : [v];
                creates.forEach((create, index) => {
                    var _a;
                    const baseName = `${varName}_create_${entry[0]}${relationField.union ? `_${refNode.name}` : ""}${index}`;
                    const nodeName = `${baseName}_node`;
                    const propertiesName = `${baseName}_relationship`;
                    const relTypeStr = `[${relationField.properties ? propertiesName : ""}:${relationField.type}]`;
                    const createAndParams = create_create_and_params_1.default({
                        context,
                        node: refNode,
                        input: create.node,
                        varName: nodeName,
                        withVars: [varName, nodeName],
                    });
                    createStrs.push(createAndParams[0]);
                    cypherParams = Object.assign(Object.assign({}, cypherParams), createAndParams[1]);
                    createStrs.push(`MERGE (${varName})${inStr}${relTypeStr}${outStr}(${nodeName})`);
                    if (relationField.properties) {
                        const relationship = context.neoSchema.relationships.find((x) => x.properties === relationField.properties);
                        const setA = create_set_relationship_properties_and_params_1.default({
                            properties: (_a = create.edge) !== null && _a !== void 0 ? _a : {},
                            varName: propertiesName,
                            relationship,
                            operation: "CREATE",
                        });
                        createStrs.push(setA[0]);
                        cypherParams = Object.assign(Object.assign({}, cypherParams), setA[1]);
                    }
                });
            });
        });
    }
    if (deleteInput) {
        const deleteAndParams = create_delete_and_params_1.default({
            context,
            node,
            deleteInput,
            varName: `${varName}_delete`,
            parentVar: varName,
            withVars: [varName],
            parameterPrefix: `${resolveTree.name}.args.delete`,
        });
        [deleteStr] = deleteAndParams;
        cypherParams = Object.assign(Object.assign({}, cypherParams), deleteAndParams[1]);
        updateArgs = Object.assign(Object.assign({}, updateArgs), (deleteStr.includes(resolveTree.name) ? { delete: deleteInput } : {}));
    }
    const projection = create_projection_and_params_1.default({
        node,
        context,
        fieldsByTypeName,
        varName,
    });
    [projStr] = projection;
    cypherParams = Object.assign(Object.assign({}, cypherParams), projection[1]);
    if ((_b = (_a = projection[2]) === null || _a === void 0 ? void 0 : _a.authValidateStrs) === null || _b === void 0 ? void 0 : _b.length) {
        projAuth = `CALL apoc.util.validate(NOT(${projection[2].authValidateStrs.join(" AND ")}), "${constants_1.AUTH_FORBIDDEN_ERROR}", [0])`;
    }
    if ((_d = (_c = projection[2]) === null || _c === void 0 ? void 0 : _c.connectionFields) === null || _d === void 0 ? void 0 : _d.length) {
        projection[2].connectionFields.forEach((connectionResolveTree) => {
            const connectionField = node.connectionFields.find((x) => x.fieldName === connectionResolveTree.name);
            const connection = create_connection_and_params_1.default({
                resolveTree: connectionResolveTree,
                field: connectionField,
                context,
                nodeVariable: varName,
            });
            connectionStrs.push(connection[0]);
            cypherParams = Object.assign(Object.assign({}, cypherParams), connection[1]);
        });
    }
    const cypher = [
        matchStr,
        whereStr,
        updateStr,
        connectStrs.join("\n"),
        disconnectStrs.join("\n"),
        createStrs.join("\n"),
        deleteStr,
        ...(connectionStrs.length || projAuth ? [`WITH ${varName}`] : []),
        ...(projAuth ? [projAuth] : []),
        ...connectionStrs,
        `RETURN ${varName} ${projStr} AS ${varName}`,
    ];
    return [
        cypher.filter(Boolean).join("\n"),
        Object.assign(Object.assign({}, cypherParams), (Object.keys(updateArgs).length ? { [resolveTree.name]: { args: updateArgs } } : {})),
    ];
}
exports.default = translateUpdate;
